File: backend/app/services/admin_service.py
"""管理后台服务：统计、用户/动态/评论/邀请码管理。"""

from __future__ import annotations

from sqlalchemy import func, or_
from sqlalchemy.orm import Session

from app.models.comments import Comment
from app.models.invite_codes import InviteCode
from app.models.likes import Like
from app.models.posts import Post
from app.models.users import User
from app.schemas.admin import (
    AdminAuthorOut,
    AdminCommentListOut,
    AdminCommentOut,
    AdminInviteListOut,
    AdminInviteOut,
    AdminPostListOut,
    AdminPostOut,
    AdminUserListOut,
    AdminUserOut,
    AdminUserUpdateIn,
    StatsOut,
)
from app.schemas.common import AppError, ErrorCode
from app.services.user_service import avatar_url_for
from app.utils.time import utcnow


def _build_admin_author(user: User) -> dict:
    is_deactivated = user.status == "deactivated"
    if is_deactivated:
        return AdminAuthorOut(
            id=user.id,
            email=user.email,
            nickname="已注销",
            avatar_url="/api/v1/avatars/default",
            is_deactivated=True,
        ).model_dump(mode="json")
    return AdminAuthorOut(
        id=user.id,
        email=user.email,
        nickname=user.nickname,
        avatar_url=avatar_url_for(user),
        is_deactivated=False,
    ).model_dump(mode="json")


def _user_to_admin_out(user: User) -> dict:
    return AdminUserOut(
        id=user.id,
        email=user.email,
        nickname=user.nickname,
        role=user.role,
        status=user.status,
        can_invite=user.can_invite,
        avatar_url=avatar_url_for(user),
        created_at=user.created_at,
        last_login_at=user.last_login_at,
    ).model_dump(mode="json")


def get_stats(db: Session) -> dict:
    user_count = db.query(func.count(User.id)).scalar() or 0
    post_count = db.query(func.count(Post.id)).filter(Post.deleted_at.is_(None)).scalar() or 0
    comment_count = (
        db.query(func.count(Comment.id)).filter(Comment.deleted_at.is_(None)).scalar() or 0
    )
    like_count = db.query(func.count(Like.id)).scalar() or 0
    invite_count = db.query(func.count(InviteCode.id)).scalar() or 0
    used_invite_count = (
        db.query(func.count(InviteCode.id)).filter(InviteCode.status == "used").scalar() or 0
    )
    return StatsOut(
        user_count=user_count,
        post_count=post_count,
        comment_count=comment_count,
        like_count=like_count,
        invite_count=invite_count,
        used_invite_count=used_invite_count,
    ).model_dump(mode="json")


def list_users(db: Session, page: int, page_size: int, search: str | None) -> dict:
    base_q = db.query(User)
    if search:
        pattern = f"%{search}%"
        base_q = base_q.filter(or_(User.email.ilike(pattern), User.nickname.ilike(pattern)))
    total = base_q.count()
    offset = (page - 1) * page_size
    users = base_q.order_by(User.id.asc()).offset(offset).limit(page_size).all()
    items = [_user_to_admin_out(u) for u in users]
    has_more = offset + page_size < total
    return AdminUserListOut(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        has_more=has_more,
    ).model_dump(mode="json")


def update_user(db: Session, user_id: int, data: AdminUserUpdateIn) -> dict:
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise AppError(ErrorCode.USER_NOT_FOUND, "User not found", 404)

    if data.restore is True:
        if user.status == "deactivated":
            user.status = "active"

    if data.status is not None:
        if user.status == "deactivated":
            raise AppError(
                ErrorCode.VALIDATION_ERROR,
                "Cannot set deactivated via admin; use restore",
                400,
            )
        user.status = data.status

    if data.can_invite is not None:
        user.can_invite = data.can_invite

    db.commit()
    db.refresh(user)
    return _user_to_admin_out(user)


def list_posts(
    db: Session,
    page: int,
    page_size: int,
    user_id: int | None,
    visibility: str | None,
) -> dict:
    base_q = db.query(Post)
    if user_id is not None:
        base_q = base_q.filter(Post.user_id == user_id)
    if visibility is not None:
        base_q = base_q.filter(Post.visibility == visibility)
    total = base_q.count()
    offset = (page - 1) * page_size
    posts = (
        base_q.order_by(Post.created_at.desc(), Post.id.desc())
        .offset(offset)
        .limit(page_size)
        .all()
    )

    if not posts:
        return AdminPostListOut(
            items=[],
            total=total,
            page=page,
            page_size=page_size,
            has_more=False,
        ).model_dump(mode="json")

    author_ids = list({p.user_id for p in posts})
    authors = {u.id: u for u in db.query(User).filter(User.id.in_(author_ids)).all()}

    post_ids = [p.id for p in posts]
    like_counts_rows = (
        db.query(Like.target_id, func.count(Like.id))
        .filter(Like.target_type == "post", Like.target_id.in_(post_ids))
        .group_by(Like.target_id)
        .all()
    )
    like_counts = {row[0]: row[1] for row in like_counts_rows}

    comment_counts_rows = (
        db.query(Comment.post_id, func.count(Comment.id))
        .filter(Comment.post_id.in_(post_ids), Comment.deleted_at.is_(None))
        .group_by(Comment.post_id)
        .all()
    )
    comment_counts = {row[0]: row[1] for row in comment_counts_rows}

    items: list[dict] = []
    for p in posts:
        author = authors.get(p.user_id)
        assert author is not None
        items.append(
            AdminPostOut(
                id=p.id,
                content=p.content,
                visibility=p.visibility,
                deleted=p.deleted_at is not None,
                author=_build_admin_author(author),
                like_count=like_counts.get(p.id, 0),
                comment_count=comment_counts.get(p.id, 0),
                created_at=p.created_at,
                deleted_at=p.deleted_at,
            ).model_dump(mode="json")
        )

    has_more = offset + page_size < total
    return AdminPostListOut(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        has_more=has_more,
    ).model_dump(mode="json")


def delete_post(db: Session, post_id: int) -> None:
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise AppError(ErrorCode.NOT_FOUND, "Post not found", 404)
    if post.deleted_at is not None:
        return
    post.deleted_at = utcnow()
    db.commit()


def list_comments(db: Session, page: int, page_size: int) -> dict:
    base_q = db.query(Comment)
    total = base_q.count()
    offset = (page - 1) * page_size
    comments = (
        base_q.order_by(Comment.created_at.desc(), Comment.id.desc())
        .offset(offset)
        .limit(page_size)
        .all()
    )

    if not comments:
        return AdminCommentListOut(
            items=[],
            total=total,
            page=page,
            page_size=page_size,
            has_more=False,
        ).model_dump(mode="json")

    author_ids = list({c.user_id for c in comments})
    authors = {u.id: u for u in db.query(User).filter(User.id.in_(author_ids)).all()}

    comment_ids = [c.id for c in comments]
    like_counts_rows = (
        db.query(Like.target_id, func.count(Like.id))
        .filter(Like.target_type == "comment", Like.target_id.in_(comment_ids))
        .group_by(Like.target_id)
        .all()
    )
    like_counts = {row[0]: row[1] for row in like_counts_rows}

    items: list[dict] = []
    for c in comments:
        author = authors.get(c.user_id)
        assert author is not None
        image_thumb_url = f"/api/v1/comments/{c.id}/media/thumb" if c.image_thumb_path else None
        items.append(
            AdminCommentOut(
                id=c.id,
                post_id=c.post_id,
                content=c.content,
                image_thumb_url=image_thumb_url,
                deleted=c.deleted_at is not None,
                author=_build_admin_author(author),
                like_count=like_counts.get(c.id, 0),
                created_at=c.created_at,
                deleted_at=c.deleted_at,
            ).model_dump(mode="json")
        )

    has_more = offset + page_size < total
    return AdminCommentListOut(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        has_more=has_more,
    ).model_dump(mode="json")


def delete_comment(db: Session, comment_id: int) -> None:
    comment = db.query(Comment).filter(Comment.id == comment_id).first()
    if not comment:
        raise AppError(ErrorCode.COMMENT_NOT_FOUND, "Comment not found", 404)
    if comment.deleted_at is not None:
        return
    comment.deleted_at = utcnow()
    db.commit()


def list_invites(db: Session, page: int, page_size: int) -> dict:
    base_q = db.query(InviteCode)
    total = base_q.count()
    offset = (page - 1) * page_size
    invites = base_q.order_by(InviteCode.created_at.desc()).offset(offset).limit(page_size).all()

    if not invites:
        return AdminInviteListOut(
            items=[],
            total=total,
            page=page,
            page_size=page_size,
            has_more=False,
        ).model_dump(mode="json")

    creator_ids = list({i.creator_id for i in invites})
    creators = {u.id: u for u in db.query(User).filter(User.id.in_(creator_ids)).all()}

    items: list[dict] = []
    for i in invites:
        creator = creators.get(i.creator_id)
        assert creator is not None
        items.append(
            AdminInviteOut(
                id=i.id,
                code=i.code,
                status=i.status,
                expires_at=i.expires_at,
                created_at=i.created_at,
                used_by_id=i.used_by_id,
                creator=_build_admin_author(creator),
            ).model_dump(mode="json")
        )

    has_more = offset + page_size < total
    return AdminInviteListOut(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        has_more=has_more,
    ).model_dump(mode="json")


def revoke_invite(db: Session, invite_id: int) -> dict:
    invite = db.query(InviteCode).filter(InviteCode.id == invite_id).first()
    if not invite:
        raise AppError(ErrorCode.INVITE_NOT_FOUND, "Invite not found", 404)
    if invite.status == "used":
        raise AppError(ErrorCode.INVITE_ALREADY_USED, "Used invite cannot be revoked", 400)
    if invite.status in ("revoked", "expired"):
        return {"message": "Invite already inactive"}
    invite.status = "revoked"
    db.commit()
    return {"message": "Invite revoked"}

File: backend/internal/service/admin_service.go
package service

import (
	"context"
	"fmt"

	"backend/internal/config"
	"backend/internal/model"
	"backend/internal/repository"
)

type AdminService struct {
	cfg         *config.Config
	repos       *repository.Repositories
	userService *UserService
	llmClient   *LLMClient
}

func NewAdminService(cfg *config.Config, repos *repository.Repositories, userService *UserService, llmClient *LLMClient) *AdminService {
	return &AdminService{
		cfg:         cfg,
		repos:       repos,
		userService: userService,
		llmClient:   llmClient,
	}
}

func (s *AdminService) GetStats(ctx context.Context) (*model.StatsOut, error) {
	userCount, err := s.repos.User.CountAll(ctx)
	if err != nil {
		return nil, err
	}
	postCount, err := s.repos.Post.CountActive(ctx)
	if err != nil {
		return nil, err
	}
	commentCount, err := s.repos.Comment.CountActive(ctx)
	if err != nil {
		return nil, err
	}
	likeCount, err := s.repos.Like.CountAll(ctx)
	if err != nil {
		return nil, err
	}
	totalInvites, usedInvites, err := s.repos.Invite.CountAll(ctx)
	if err != nil {
		return nil, err
	}

	return &model.StatsOut{
		UserCount:       userCount,
		PostCount:       postCount,
		CommentCount:    commentCount,
		LikeCount:       likeCount,
		InviteCount:     totalInvites,
		UsedInviteCount: usedInvites,
	}, nil
}

func (s *AdminService) ListUsers(ctx context.Context, page, pageSize int, search string) (*model.AdminUserListOut, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	} else if pageSize > 100 {
		pageSize = 100
	}

	offset := (page - 1) * pageSize
	users, total, err := s.repos.User.ListAdmin(ctx, search, offset, pageSize)
	if err != nil {
		return nil, err
	}

	var items []model.AdminUserOut
	for _, u := range users {
		items = append(items, model.AdminUserOut{
			ID:          u.ID,
			Email:       u.Email,
			Nickname:    u.Nickname,
			Role:        u.Role,
			Status:      u.Status,
			CanInvite:   u.CanInvite,
			AvatarURL:   s.userService.AvatarURLFor(u),
			CreatedAt:   u.CreatedAt,
			LastLoginAt: u.LastLoginAt,
		})
	}
	if items == nil {
		items = []model.AdminUserOut{}
	}

	hasMore := offset+pageSize < total
	return &model.AdminUserListOut{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
		HasMore:  hasMore,
	}, nil
}

func (s *AdminService) UpdateUser(ctx context.Context, userID int, data model.AdminUserUpdateIn) (*model.AdminUserOut, error) {
	user, err := s.repos.User.GetByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, model.NewAppError(model.ErrUserNotFound, "用户不存在", 404)
	}

	if data.Restore != nil && *data.Restore {
		if user.Status == "deactivated" {
			user.Status = "active"
		}
	}

	if data.Status != nil {
		if user.Status == "deactivated" {
			return nil, model.NewAppError(model.ErrValidationError, "无法直接修改已注销用户的状态，请使用恢复操作", 400)
		}
		user.Status = *data.Status
	}

	if data.CanInvite != nil {
		user.CanInvite = *data.CanInvite
	}

	if err := s.repos.User.Update(ctx, user); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "更新用户状态失败", 500)
	}

	return &model.AdminUserOut{
		ID:          user.ID,
		Email:       user.Email,
		Nickname:    user.Nickname,
		Role:        user.Role,
		Status:      user.Status,
		CanInvite:   user.CanInvite,
		AvatarURL:   s.userService.AvatarURLFor(user),
		CreatedAt:   user.CreatedAt,
		LastLoginAt: user.LastLoginAt,
	}, nil
}

func (s *AdminService) ListPosts(ctx context.Context, page, pageSize int, userID *int, visibility *string) (*model.AdminPostListOut, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	} else if pageSize > 100 {
		pageSize = 100
	}

	offset := (page - 1) * pageSize
	posts, total, err := s.repos.Post.ListAdmin(ctx, userID, visibility, offset, pageSize)
	if err != nil {
		return nil, err
	}

	if len(posts) == 0 {
		return &model.AdminPostListOut{
			Items:    []model.AdminPostOut{},
			Total:    total,
			Page:     page,
			PageSize: pageSize,
			HasMore:  false,
		}, nil
	}

	postIDs := make([]int, len(posts))
	authorIDs := make(map[int]bool)
	for i, p := range posts {
		postIDs[i] = p.ID
		authorIDs[p.UserID] = true
	}

	likeCounts, _ := s.repos.Post.BatchPreloadLikeCounts(ctx, postIDs)
	commentCounts, _ := s.repos.Post.BatchPreloadCommentCounts(ctx, postIDs)

	var uids []int
	for uid := range authorIDs {
		uids = append(uids, uid)
	}
	usersMap, _ := s.repos.User.GetByIDs(ctx, uids)

	var items []model.AdminPostOut
	for _, p := range posts {
		author := usersMap[p.UserID]
		authorOut := s.buildAdminAuthorOut(author)

		items = append(items, model.AdminPostOut{
			ID:           p.ID,
			Content:      p.Content,
			Visibility:   p.Visibility,
			Deleted:      p.DeletedAt != nil,
			Author:       authorOut,
			LikeCount:    likeCounts[p.ID],
			CommentCount: commentCounts[p.ID],
			CreatedAt:    p.CreatedAt,
			DeletedAt:    p.DeletedAt,
		})
	}

	hasMore := offset+pageSize < total
	return &model.AdminPostListOut{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
		HasMore:  hasMore,
	}, nil
}

func (s *AdminService) DeletePost(ctx context.Context, postID int) error {
	post, err := s.repos.Post.GetByIDIncludeDeleted(ctx, postID)
	if err != nil {
		return err
	}
	if post == nil {
		return model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}
	if post.DeletedAt != nil {
		return nil
	}
	return s.repos.Post.SoftDelete(ctx, postID)
}

func (s *AdminService) ListComments(ctx context.Context, page, pageSize int) (*model.AdminCommentListOut, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	} else if pageSize > 100 {
		pageSize = 100
	}

	offset := (page - 1) * pageSize
	comments, total, err := s.repos.Comment.ListAdmin(ctx, offset, pageSize)
	if err != nil {
		return nil, err
	}

	if len(comments) == 0 {
		return &model.AdminCommentListOut{
			Items:    []model.AdminCommentOut{},
			Total:    total,
			Page:     page,
			PageSize: pageSize,
			HasMore:  false,
		}, nil
	}

	commentIDs := make([]int, len(comments))
	authorIDs := make(map[int]bool)
	for i, c := range comments {
		commentIDs[i] = c.ID
		authorIDs[c.UserID] = true
	}

	likeCounts, _ := s.repos.Comment.BatchPreloadLikeCounts(ctx, commentIDs)

	var uids []int
	for uid := range authorIDs {
		uids = append(uids, uid)
	}
	usersMap, _ := s.repos.User.GetByIDs(ctx, uids)

	var items []model.AdminCommentOut
	for _, c := range comments {
		author := usersMap[c.UserID]
		authorOut := s.buildAdminAuthorOut(author)

		var thumbURL *string
		if c.ImageThumbPath != nil {
			u := fmt.Sprintf("/api/v1/comments/%d/media/thumb", c.ID)
			thumbURL = &u
		}

		items = append(items, model.AdminCommentOut{
			ID:            c.ID,
			PostID:        c.PostID,
			Content:       c.Content,
			ImageThumbURL: thumbURL,
			Deleted:       c.DeletedAt != nil,
			Author:        authorOut,
			LikeCount:     likeCounts[c.ID],
			CreatedAt:     c.CreatedAt,
			DeletedAt:     c.DeletedAt,
		})
	}

	hasMore := offset+pageSize < total
	return &model.AdminCommentListOut{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
		HasMore:  hasMore,
	}, nil
}

func (s *AdminService) DeleteComment(ctx context.Context, commentID int) error {
	comment, err := s.repos.Comment.GetByIDIncludeDeleted(ctx, commentID)
	if err != nil {
		return err
	}
	if comment == nil {
		return model.NewAppError(model.ErrCommentNotFound, "评论不存在", 404)
	}
	if comment.DeletedAt != nil {
		return nil
	}
	return s.repos.Comment.SoftDelete(ctx, commentID)
}

func (s *AdminService) ListInvites(ctx context.Context, page, pageSize int) (*model.AdminInviteListOut, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	} else if pageSize > 100 {
		pageSize = 100
	}

	offset := (page - 1) * pageSize
	invites, total, err := s.repos.Invite.ListAdmin(ctx, offset, pageSize)
	if err != nil {
		return nil, err
	}

	var creatorIDs []int
	for _, inv := range invites {
		creatorIDs = append(creatorIDs, inv.CreatorID)
	}
	usersMap, _ := s.repos.User.GetByIDs(ctx, creatorIDs)

	var items []model.AdminInviteOut
	for _, inv := range invites {
		creator := usersMap[inv.CreatorID]
		items = append(items, model.AdminInviteOut{
			ID:        inv.ID,
			Code:      inv.Code,
			Status:    inv.Status,
			ExpiresAt: inv.ExpiresAt,
			CreatedAt: inv.CreatedAt,
			UsedByID:  inv.UsedByID,
			Creator:   s.buildAdminAuthorOut(creator),
		})
	}
	if items == nil {
		items = []model.AdminInviteOut{}
	}

	hasMore := offset+pageSize < total
	return &model.AdminInviteListOut{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
		HasMore:  hasMore,
	}, nil
}

func (s *AdminService) RevokeInvite(ctx context.Context, inviteID int) (*model.InviteActionOut, error) {
	inv, err := s.repos.Invite.GetByID(ctx, inviteID)
	if err != nil {
		return nil, err
	}
	if inv == nil {
		return nil, model.NewAppError(model.ErrInviteNotFound, "邀请码不存在", 404)
	}
	if inv.Status == "used" {
		return nil, model.NewAppError(model.ErrInviteAlreadyUsed, "已使用的邀请码无法失效", 400)
	}
	if inv.Status == "revoked" || inv.Status == "expired" {
		return &model.InviteActionOut{Message: "邀请码已失效"}, nil
	}

	if err := s.repos.Invite.Revoke(ctx, inviteID); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "失效邀请码失败", 500)
	}
	return &model.InviteActionOut{Message: "邀请码已失效"}, nil
}

func (s *AdminService) GetLLMConfig(ctx context.Context) (*model.LLMConfigOut, error) {
	st, err := s.repos.System.GetStatus(ctx)
	if err != nil {
		return nil, err
	}
	if st == nil {
		return nil, model.NewAppError(model.ErrNotFound, "系统状态未初始化", 500)
	}

	return &model.LLMConfigOut{
		BaseURL:   st.LLMBaseURL,
		Model:     st.LLMModel,
		Timeout:   st.LLMTimeout,
		MaxTokens: st.LLMMaxTokens,
		HasAPIKey: st.LLMAPIKey != "",
	}, nil
}

func (s *AdminService) UpdateLLMConfig(ctx context.Context, data model.LLMConfigUpdateIn) (*model.LLMConfigOut, error) {
	st, err := s.repos.System.UpdateLLMConfig(ctx, data.BaseURL, data.APIKey, data.Model, data.Timeout, data.MaxTokens)
	if err != nil {
		return nil, model.NewAppError(model.ErrInternal, "更新大模型配置失败", 500)
	}

	return &model.LLMConfigOut{
		BaseURL:   st.LLMBaseURL,
		Model:     st.LLMModel,
		Timeout:   st.LLMTimeout,
		MaxTokens: st.LLMMaxTokens,
		HasAPIKey: st.LLMAPIKey != "",
	}, nil
}

func (s *AdminService) TestLLMConfig(ctx context.Context, data model.LLMConfigTestIn) (*model.LLMTestOut, error) {
	st, err := s.repos.System.GetStatus(ctx)
	if err != nil || st == nil {
		return nil, model.NewAppError(model.ErrNotFound, "系统状态未初始化", 500)
	}

	baseURL := st.LLMBaseURL
	if data.BaseURL != nil && *data.BaseURL != "" {
		baseURL = *data.BaseURL
	}
	apiKey := st.LLMAPIKey
	if data.APIKey != nil && *data.APIKey != "" {
		apiKey = *data.APIKey
	}
	modelName := st.LLMModel
	if data.Model != nil && *data.Model != "" {
		modelName = *data.Model
	}
	timeout := st.LLMTimeout
	if data.Timeout != nil && *data.Timeout > 0 {
		timeout = *data.Timeout
	}

	reply, err := s.llmClient.TestLLM(ctx, baseURL, apiKey, modelName, timeout)
	if err != nil {
		return &model.LLMTestOut{
			Success: false,
			Message: fmt.Sprintf("测试失败：%v", err),
		}, nil
	}

	return &model.LLMTestOut{
		Success: true,
		Message: fmt.Sprintf("测试成功，模型回复：%s", reply),
	}, nil
}

func (s *AdminService) buildAdminAuthorOut(u *model.User) model.AdminAuthorOut {
	if u == nil {
		return model.AdminAuthorOut{
			ID:            0,
			Email:         "",
			Nickname:      "未知",
			AvatarURL:     "/api/v1/avatars/default",
			IsDeactivated: false,
		}
	}
	if u.Status == "deactivated" {
		return model.AdminAuthorOut{
			ID:            u.ID,
			Email:         u.Email,
			Nickname:      "已注销",
			AvatarURL:     "/api/v1/avatars/default",
			IsDeactivated: true,
		}
	}
	avatarURL := "/api/v1/avatars/default"
	if u.AvatarPath != nil && *u.AvatarPath != "" {
		avatarURL = fmt.Sprintf("/api/v1/avatars/%d", u.ID)
	}
	return model.AdminAuthorOut{
		ID:            u.ID,
		Email:         u.Email,
		Nickname:      u.Nickname,
		AvatarURL:     avatarURL,
		IsDeactivated: false,
	}
}
